# <font color=blue> Naive Bayes 

### <font color=byellow> ใช้ไฟล์ ContactLens.csv ไฟล์ ประกอบด้วยคุณลักษณะดังนี้
- ### <font color=blue> Age (กลุ่มอายุคนไข้) : young, pre-presbyopic, presbyopic (สายตาผู้สูงอายุ)
- ### <font color=blue> Prescript (ลักษณะการมองเห็น): myope(สายตาสั้น), hypermetrope(สายตายาว)
- ### <font color=blue> Astigmatism (สายตาเอียง): no, yes
- ### <font color=blue> TearProdRate (น้ำตา) : reduced, normal
- ### <font color=blue> ContactLens มี 3 ประเภท คือ none, soft, hard

### <font color=bgreen> ลุงตู่มาตรวจสายตาเพื่อจะได้ทราบว่าต้องใส่ contact lens หรือไม่รายละเอียดมีดังนี้
- ### <font color=blue> ลุงตู่ => young, myope, yes, reduced

In [1]:
import pandas as pd
import numpy as np

In [2]:
#เขียนโปรแกรมโดยใช้วิธี naive bayes เพื่อช่วยให้คำแนะนำเบื้องต้นว่าลงตู่น่าจะใส่ contact lens ประเภทไหน 
data = pd.read_csv("ContactLens.csv")
print(data)

               Age     Prescript Astigmatism TearProdRate ContactLens
0            young         myope          no      reduced        none
1            young         myope          no       normal        soft
2            young         myope         yes      reduced        none
3            young         myope         yes       normal        hard
4            young  hypermetrope          no      reduced        none
5            young  hypermetrope          no       normal        soft
6            young  hypermetrope         yes      reduced        none
7            young  hypermetrope         yes       normal        hard
8   pre-presbyopic         myope          no      reduced        none
9   pre-presbyopic         myope          no       normal        soft
10  pre-presbyopic         myope         yes      reduced        none
11  pre-presbyopic         myope         yes       normal        hard
12  pre-presbyopic  hypermetrope          no      reduced        none
13  pre-presbyopic  

In [7]:
new = pd.DataFrame()

new['Age'] = ['young']
new['Prescript'] = ['myope']
new['Astigmatism'] = ['yes']
new['TearProdRate'] = ['reduced']

new

Age Prescript Astigmatism TearProdRate
0  young     myope         yes      reduced

In [9]:
n_none = len(data[data['ContactLens'] == 'none'])
n_soft = len(data[data['ContactLens'] == 'soft'])
n_hard = len(data[data['ContactLens'] == 'hard'])

total_use = len(data['ContactLens'])
print(n_none, n_soft,n_hard, total_use)

15 5 4 24


In [10]:
n_young_none = len(data[(data['Age'] == 'young') & (data['ContactLens'] == 'none')])
n_myope_none = len(data[(data['Prescript'] == 'myope') & (data['ContactLens'] == 'none')])
n_yes_none = len(data[(data['Astigmatism'] == 'yes') & (data['ContactLens'] == 'none')])
n_reduced_none = len(data[(data['TearProdRate'] == 'reduced') & (data['ContactLens'] == 'none')])

In [11]:
n_young_soft = len(data[(data['Age'] == 'young') & (data['ContactLens'] == 'soft')])
n_myope_soft = len(data[(data['Age'] == 'young') & (data['ContactLens'] == 'soft')])
n_yes_soft = len(data[(data['Age'] == 'young') & (data['ContactLens'] == 'soft')])
n_reduced_soft = len(data[(data['Age'] == 'young') & (data['ContactLens'] == 'soft')])

In [12]:
n_young_hard = len(data[(data['Age'] == 'young') & (data['ContactLens'] == 'hard')])
n_myope_hard = len(data[(data['Age'] == 'young') & (data['ContactLens'] == 'hard')])
n_yes_hard = len(data[(data['Age'] == 'young') & (data['ContactLens'] == 'hard')])
n_reduced_hard = len(data[(data['Age'] == 'young') & (data['ContactLens'] == 'hard')])

In [18]:
p_none = n_none/total_use
p_soft = n_soft/total_use
p_hard = n_hard/total_use

p_young_none = n_young_none/n_none
p_myope_none = n_myope_none/n_none
p_yes_none = n_yes_none/n_none
p_reduced_none = n_reduced_none/n_none


In [19]:
p_young_soft = n_young_soft/n_soft
p_myope_soft = n_myope_soft/n_soft
p_yes_soft = n_yes_soft/n_soft
p_reduced_soft = n_reduced_soft/n_soft

In [20]:
p_young_hard = n_young_soft/n_hard
p_myope_hard = n_myope_soft/n_hard
p_yes_hard = n_yes_soft/n_hard
p_reduced_hard = n_reduced_soft/n_hard

In [24]:
p_young_myope_yes_reduced_none = p_young_none* p_myope_none * p_yes_none * p_reduced_none * p_none
p_young_myope_yes_reduced_soft = p_young_soft* p_myope_soft * p_yes_soft * p_reduced_soft * p_soft
p_young_myope_yes_reduced_hard = p_young_hard* p_myope_hard * p_yes_hard * p_reduced_soft * p_hard

print('p_young_myope_yes_reduced_none =' , "%.4f" %p_young_myope_yes_reduced_none)
print('p_young_myope_yes_reduced_soft =' , "%.4f" %p_young_myope_yes_reduced_soft)
print('p_young_myope_yes_reduced_hard =' , "%.4f" %p_young_myope_yes_reduced_hard)


p_young_myope_yes_reduced_none = 0.0332
p_young_myope_yes_reduced_soft = 0.0053
p_young_myope_yes_reduced_hard = 0.0083


### <font color=green> ตอบช่องนี้ว่าลุงตู่น่าจะใส้ contact lens แบบไหน ให้เหตุผลอธิบาย   ค่าP(ContactLens = none|new) = 0.0332 มีค่ามากกว่า P(ContactLens = soft|new) = 0.0053 และ P(ContactLens = hard|new) = 0.0083 ดังนั้น ลุงตู่จึงไม่ใส่ ContactLens